### PR TITLE
 Buffer a ClientHello with a cookie received via DTLSv1_listen

### LINKED ISSUE #6934: DTLSv1_listen() regression on 1.1.x
The rewrite of DTLSv1_listen() makes multi client DTLS server impossible on OpenSSL 1.1.x.
DTLSv1_listen() peeks into the kernel queue, but does not consume the client hello. Therefore
the common approach, creating a new socket for the client and off loading it, is doomed to failure since the hello is pending and the application logic thinks a new client approached.

This PR seems to address the regression but sadly didn't get much attention yet:
https://github.com/openssl/openssl/pull/5024

For more details see my mail:
https://mta.openssl.org/pipermail/openssl-users/2018-August/008498.html

### PR DESCRIPTION
**Pull request by mattcaswell**
Previously when a ClientHello arrives with a valid cookie using `DTLSv1_listen()` we only "peeked" at the message and left it on the underlying fd. This works fine for single threaded applications but for multi-threaded apps this does not work since the fd is typically reused for the server thread, while a new fd is created and connected for the client. By "peeking" we leave the message on the server fd, and consequently we think we've received another valid ClientHello and so we create yet another
fd for the client, and so on until we run out of fds.

In this new approach we remove the ClientHello and buffer it in the SSL object.

The buffering approach requires the record data to be in the read SSL3_BUFFER object. Previously we were using the `init_buf`. It makes more sense for it to be in the SSL3_BUFFER anyway, so we swap to using these buffers instead.

Fixes #6934

### CODE REVIEW

File: ssl/d1_lib.c [hunk 1]
```diff
@@ -500,14 +493,26 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
...
+     * Using SSL3_RT_HEADER_LENGTH here instead of DTLS1_RT_HEADER_LENGTH for
+     * consistency with ssl3_read_n. In practice it should make no difference
+     * for sensible values of SSL3_ALIGN_PAYLOAD because the difference between
+     * SSL3_RT_HEADER_LENGTH and DTLS1_RT_HEADER_LENGTH is exactly 8
+     */
+    align = (size_t)buf + SSL3_RT_HEADER_LENGTH;
+    align = SSL3_ALIGN_PAYLOAD - 1 - ((align - 1) % SSL3_ALIGN_PAYLOAD);
+# endif
+#endif
+    buf += align;
```
Comment (richardweinberger) [inline ssl/d1_lib.c]: Isn't align uninitialized if !defined(SSL3_ALIGN_PAYLOAD)?

File: ssl/d1_lib.c [hunk 2]
```diff
@@ -475,13 +475,6 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
         return -1;
     }
 
-    /*
-     * We only peek at incoming ClientHello's until we're sure we are going to
-     * to respond with a HelloVerifyRequest. If its a ClientHello with a valid
-     * cookie then we leave it in the BIO for accept to handle.
-     */
-    BIO_ctrl(SSL_get_rbio(s), BIO_CTRL_DGRAM_SET_PEEK_MODE, 1, NULL);
```
Comment (richardweinberger) [inline ssl/d1_lib.c]: Since you remove all users of BIO_CTRL_DGRAM_SET_PEEK_MODE, should we get rid of it too?
Though, I don't whether BIO_CTRL_DGRAM_SET_PEEK_MODE is part of the public visiable ABI.

File: ssl/d1_lib.c [hunk 3]
```diff
@@ -495,34 +494,19 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
...
-         * Technically a ClientHello could be SSL3_RT_MAX_PLAIN_LENGTH
-         * + DTLS1_RT_HEADER_LENGTH bytes long. Normally init_buf does not store
-         * the record header as well, but we do here. We've set up init_buf to
-         * be the standard size for simplicity. In practice we shouldn't ever
-         * receive a ClientHello as long as this. If we do it will get dropped
-         * in the record length check below.
-         */
-        n = BIO_read(rbio, buf, SSL3_RT_MAX_PLAIN_LENGTH);
+        n = BIO_read(rbio, buf, SSL3_RT_MAX_PLAIN_LENGTH
+                                + DTLS1_RT_HEADER_LENGTH);
```
Comment (richardweinberger) [inline ssl/d1_lib.c]: Hmm, I'm a little confused. Isn't this a unrelated change?
Why do we read now + DTLS1_RT_HEADER_LENGTH bytes?

File: ssl/d1_lib.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/d1_lib.c]: It's part of the public ABI so cannot be removed

File: ssl/d1_lib.c (hunk 1 above)
Comment (mattcaswell) [reply ssl/d1_lib.c]: Yes - good catch

File: ssl/d1_lib.c (hunk 3 above)
Comment (mattcaswell) [reply ssl/d1_lib.c]: In the comment I removed you'll see that we used a buffer size slightly less that we would ideally like because of the use of `init_buf`. Now we we are using the record layer rbuf instead which is a bit bigger, and big enough to hold a full length record. Therefore we might as well use it to read the max size that is theoretically possible.

Comment (richardweinberger) [reply ssl/d1_lib.c]: Ah ok. the new buffer can deal with that. Makes sense now!

**Comment by mattcaswell** [comment]
Pushed a fixup commit to address the uninitialised `align` problem.

Ping @openssl/committers.

**Comment by kaduk** [comment]
[I have some desire to review this, but this week is pretty bad for me, so if you get enough other reviews that you are satisfied, don't feel like you need to wait for me.]

**Comment by mcr** [comment]
Thank you for offering this fix.  I think it will work, and I'll try to get to see if I can adapt my ruby-openssl changes to use this API.

File: ssl/d1_lib.c [hunk 4]
```diff
@@ -865,15 +848,12 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
     if (BIO_dgram_get_peer(rbio, client) <= 0)
         BIO_ADDR_clear(client);
 
+    if (!dtls_buffer_listen_record(s, reclen, seq, align))
```
Comment (kaduk) [inline ssl/d1_lib.c]: I understand that this is an internal API and we can change it later if we need to, but it might be worth thinking about pulling up the processed_rcds reference to this call, to be clear that we're adding to the "already-processed" queue.

File: ssl/d1_lib.c [hunk 5]
```diff
@@ -495,35 +487,32 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
...
-            return -1;
-        }
-
-        if (!BUF_MEM_grow(bufm, SSL3_RT_MAX_PLAIN_LENGTH)) {
-            BUF_MEM_free(bufm);
-            SSLerr(SSL_F_DTLSV1_LISTEN, ERR_R_MALLOC_FAILURE);
-            return -1;
-        }
-        s->init_buf = bufm;
+    if (!ssl3_setup_buffers(s)) {
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/d1_lib.c]: Right. `ssl3_setup_buffers()` doesn't do anything if the buffers have already been set up, so no matter how many times we go around the loop calling DTLSv1_listen() we still only allocate the buffers once per cookie-bearing-ClientHello.

File: ssl/record/ssl3_record.c [hunk 6]
```diff
@@ -2030,3 +2030,28 @@ int dtls1_get_record(SSL *s)
...
+{
+    SSL3_RECORD *rr;
+
+    rr = RECORD_LAYER_get_rrec(&s->rlayer);
+    memset(rr, 0, sizeof(SSL3_RECORD));
+
+    rr->length = len;
+    rr->type = SSL3_RT_HANDSHAKE;
+    memcpy(rr->seq_num, seq, sizeof(rr->seq_num));
+    rr->off = off;
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/record/ssl3_record.c]: I only discovered we needed to do it because it crashed with an alignment problem without it ;-)

File: ssl/d1_lib.c [hunk 7]
```diff
@@ -732,7 +711,11 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
                                                                : s->version;
 
             /* Construct the record and message headers */
-            if (!WPACKET_init(&wpkt, s->init_buf)
+            if (!WPACKET_init_static_len(&wpkt,
+                                         wbuf,
+                                         SSL3_RT_MAX_PLAIN_LENGTH
+                                         + DTLS1_RT_HEADER_LENGTH,
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/d1_lib.c]: I added a fixup commit to use `ssl_get_max_send_fragment()` instead of `SSL3_RT_MAX_PLAIN_LENGTH`. We still need to add the header length on though.

File: ssl/d1_lib.c [hunk 8]
```diff
@@ -495,35 +487,32 @@ int DTLSv1_listen(SSL *s, BIO_ADDR *client)
...
+    wbuf = RECORD_LAYER_get_wbuf(&s->rlayer)[0].buf;
+#if defined(SSL3_ALIGN_PAYLOAD)
+# if SSL3_ALIGN_PAYLOAD != 0
+    /*
+     * Using SSL3_RT_HEADER_LENGTH here instead of DTLS1_RT_HEADER_LENGTH for
+     * consistency with ssl3_read_n. In practice it should make no difference
+     * for sensible values of SSL3_ALIGN_PAYLOAD because the difference between
+     * SSL3_RT_HEADER_LENGTH and DTLS1_RT_HEADER_LENGTH is exactly 8
+     */
+    align = (size_t)buf + SSL3_RT_HEADER_LENGTH;
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply ssl/d1_lib.c]: Because I want the alignment calculation used to be the same as the alignment calculation used everywhere else in DTLS. Since `ssl3_read_n` is used for all other reads in DTLS it makes sense to me to be consistent with it.

File: ssl/d1_lib.c (hunk 4 above)
Comment (mattcaswell) [reply ssl/d1_lib.c]: It seems a little overkill to me to add it as a parameter. But I did add a comment before this line stating which queue is used.

**Comment by mattcaswell** [comment]
Fix up commit pushed.

File: ssl/record/ssl3_record.c [hunk 9]
```diff
@@ -2030,3 +2030,28 @@ int dtls1_get_record(SSL *s)
...
+
+    s->rlayer.packet = RECORD_LAYER_get_rbuf(&s->rlayer)->buf;
```
Comment (kaduk) [inline ssl/record/ssl3_record.c]: This function (dtls_buffer_listen_record()) is the part of this PR that I am least confident about.  Everything else looks good, but I want to double-check the memory semantics here.
We're setting the rlayer.packet to point to the rbuf's storage, and then calling dtls1_buffer_record().  This could be problematic if the "buffering" was just keeping a pointer reference, as the rbuf could get reallocated, the contents overwritten, etc.  But it looks like dtls1_buffer_record() is internally giong to do a memcpy to the DTLS1_RECORD_DATA's local storage and clear out the ```s->rlayer.packet``` field, avoiding the risk of any memory reference extending past the scope of this function.
So I tentatively think it all checks out, but please confirm that I'm understanding properly before merging.

Comment (mattcaswell) [reply ssl/record/ssl3_record.c]: The SSL3_BUFFER structure contains meta data about the data buffer , as well as a pointer to the malloc'd data buffer itself. "packet" points into the data buffer. `dtls_buffer_record` takes a copy of the entire SSL3_BUFFER structure - which includes the pointer to the data buffer - and then resets the whole thing to 0. It also takes a copy of "packet" - with the original also zero'd afterwards. So the original data buffer remain "active", and allocated (it isn't copied anywhere) but there are only two things that point into it: the copied SSL3_BUFFER structure that is pushed onto the queue, as well as the copy of "packet" which is also pushed on the queue as part of the same item. Whilst on the queue there is no possibility of the data buffer being reallocated.

Hopefully that helps?

Comment (kaduk) [reply ssl/record/ssl3_record.c]: It does, thanks.

**Comment by mattcaswell** [comment]
Pushed to master and 1.1.1

**Comment by YanAnzouyijun** [comment]
@mattcaswell  hello,i hava a question.
in 1.1.0g,i found DTLSv1_listen function,also hava the same question, for multi-threaded apps this does not work,am I right?
thanks.

**Comment by mattcaswell** [comment]
> in 1.1.0g,i found DTLSv1_listen function,also hava the same question, for multi-threaded apps this does not work,am I right?

@YanAnzouyijun The issue is likely to affect 1.1.0 as well, but that version is only receiving security updates - and this bug fix does not qualify. Please upgrade to 1.1.1.

**Comment by YanAnzouyijun** [comment]
@mattcaswell ,hello, i have  two questions.
could you tell me the usage of the function:dtls_buffer_listen_record ?
Also，if i can simply transplant the bug code you submit to openssl 1.1.0g ？(https://github.com/levitte/openssl/commit/fd6987529152bb9556fb2ac6c341e2e016d385f3)
because our project has decided to use the openssl 1.1.0g version. unfortunately.
I have transplanted the code to the 1.1.0g. but unfortuately, the project had coredumped in  
/* Close message body */
 || !WPACKET_close(&wpkt)

**Comment by mattcaswell** [comment]
> could you tell me how to avoid the dtls bug in 1.1.0g or how can i fix the bug in openssl 1.1.0g ？
thanks.

I'd strongly recommend against trying to copy the code from 1.1.1 to 1.1.0. It's a non-trivial exercise and you could open yourself up to security issues if it's not done right. You really should upgrade to 1.1.1.

**Comment by YanAnzouyijun** [comment]
thank you for your reply，we will update the version to 1.1.1
best regards
